rpmio/macro.c

```diff
@@ -1197,17 +1214,106 @@ static void doTrace(MacroBuf mb, int chkexist, int negate,
 static const char *setNegateAndCheck(const char *str, int *pnegate, int *pchkexist) {
 
     *pnegate = 0;
-    *pchkexist = 0;
+    *pchkexist = CHK_NO;
     while ((*str == '?') || (*str == '!')) {
 	if (*str == '!')
 	    *pnegate = !*pnegate;
 	else
-	    (*pchkexist)++;
+	    *pchkexist = CHK_BASIC;
 	str++;
     }
     return str;
 }
 
+static void partsInit(macroPartition *parts)
+{
+    parts->f = parts->fe = NULL;
+    parts->g = parts->ge = NULL;
+    parts->h = parts->he = NULL;
+    parts->fn = parts->gn = parts->hn = 0;
+    return;
+}
+
+static void setPartsSize(macroPartition *parts)
+{
+    parts->fn = parts->fe - parts->f;
+    parts->gn = parts->ge - parts->g;
+    parts->hn = parts->he - parts->h;
+    return;
+}
+
+/**
+ * Skip spaces before the input string
+ * @param str		string that starts after the ..
+ * @
+ */
+static const char *findParameterEnd(const char *str) {
+    char c;
+
+    str--;
+    SKIPBLANKREVERSE(str, c);
+    str++;
+    return str;
+}
+
+/**
+ * Checking and parsing triple operator for conditional shortcut
+ * @param source	string to expand
+ * @param sourceEnd	pointer after the valid string to expand
+ * @param parts		macro partitioned to the three possible parts
+ * @return		parsed triple condition operator macro
+ */
+static macroPartition *testTripleCondition(const char *source,
+		const char *sourceEnd, macroPartition *parts)
+{
+    char c;
+
+    parts->f = source;
+    SKIPBLANK(parts->f, c);
+
+    /* the first parameter must start by { */
+    if (parts->f[0] == '{') {
+	parts->f++;
+	/* the first parameter ends by } */
+	if ((parts->fe = matchchar(parts->f, '{', '}')) == NULL)
+	    return NULL;
+	SKIPBLANK(parts->f, c);
+	parts->g = parts->fe + 1;
+	parts->fe = findParameterEnd(parts->fe);
+    } else {
+	return NULL;
+    }
+    SKIPBLANK(parts->g, c);
+
+    /* after %{? {macroo} must be ':' */
+    if (parts->g[0] == ':') {
+	parts->g++;
+	SKIPBLANK(parts->g,c);
+	parts->ge = parts->g;
+	while ((parts->ge[0] != 0) && (parts->ge[0] != ':')) {
+	    parts->ge++;
+	    if (parts->ge[0] == '{') {
+		if ((parts->ge = matchchar(parts->ge++, '{', '}')) == NULL)
+		    return NULL;
+	    }
+	}
+    } else {
+	return NULL;
+    }
+
+    /* be the trhird part starting by ':' is optional */
+    parts->h = parts->ge;
+    parts->ge = findParameterEnd(parts->h[0] == ':' ? parts->ge : --parts->ge);
```

Comment:
Never use --/++ side-effects like that. You're even assigning to the same variable so it should be just
```? parts->ge : parts->ge -1```